File: internal/wal3/catalog.go
package wal3

import (
	"bufio"
	"bytes"
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
)

// CatRecordID values this decoder understands (the binary catalog's core
// record-type partition; bit 15 marks Enterprise types).
const (
	CatCreateDatabase     uint16 = 4
	CatSoftDeleteDatabase uint16 = 5
	CatCreateTable        uint16 = 6
	CatSoftDeleteTable    uint16 = 7
	CatAddColumns         uint16 = 8
	CatHardDeleteDatabase uint16 = 22
	CatHardDeleteTable    uint16 = 23
)

// CatRecord is one decoded catalog record. Skip is true for record types the
// migration does not need (tokens, caches, triggers, nodes, ...); Record is
// then nil.
type CatRecord struct {
	ID     uint16          `json:"id"`
	Skip   bool            `json:"skip"`
	Record json.RawMessage `json:"record"`
}

// CatCreateDatabaseRec mirrors the upstream CreateDatabase record body.
type CatCreateDatabaseRec struct {
	DatabaseID   uint32 `json:"database_id"`
	DatabaseName string `json:"database_name"`
}

// CatCreateTableRec mirrors the upstream CreateTable record body.
type CatCreateTableRec struct {
	DatabaseID   uint32 `json:"database_id"`
	DatabaseName string `json:"database_name"`
	TableName    string `json:"table_name"`
	TableID      uint32 `json:"table_id"`
}

// CatColumnDef is one entry of an AddColumns record: the externally tagged
// ColumnDefinition enum — exactly one variant set.
type CatColumnDef struct {
	Timestamp *struct {
		ColumnID *uint16 `json:"column_id"`
		Name     string  `json:"name"`
	} `json:"Timestamp"`
	Tag *struct {
		ID       uint16  `json:"id"`
		ColumnID *uint16 `json:"column_id"`
		Name     string  `json:"name"`
	} `json:"Tag"`
	Field *struct {
		ColumnID *uint16 `json:"column_id"`
		Name     string  `json:"name"`
	} `json:"Field"`
}

// CatAddColumnsRec mirrors the upstream AddColumns record body.
type CatAddColumnsRec struct {
	DatabaseID uint32         `json:"database_id"`
	TableID    uint32         `json:"table_id"`
	Columns    []CatColumnDef `json:"columns"`
}

// CatDeleteRec covers the soft/hard delete bodies. Upstream names the
// database id field "database_id" on soft deletes and "db_id" on hard
// deletes — pick by record id (both can legitimately be 0: _internal).
type CatDeleteRec struct {
	DatabaseID uint32 `json:"database_id"`
	DBID       uint32 `json:"db_id"`
	TableID    uint32 `json:"table_id"`
}

// DecodeCatalogRecords decodes a stream of raw catalog record bodies. The
// input is [u16 LE id][u32 LE length][body] per record, in replay order (the
// caller strips file and record-header framing). Output preserves order.
func (d *Decoder) DecodeCatalogRecords(ctx context.Context, stream []byte) ([]CatRecord, error) {
	if len(stream) == 0 {
		return nil, nil
	}
	stdout, err := d.run(ctx, stream, "catalog")
	if err != nil {
		return nil, err
	}
	var out []CatRecord
	sc := bufio.NewScanner(bytes.NewReader(stdout))
	sc.Buffer(make([]byte, 0, 1<<20), 1<<26)
	for sc.Scan() {
		var r CatRecord
		if err := json.Unmarshal(sc.Bytes(), &r); err != nil {
			return nil, fmt.Errorf("parse catalog decoder output: %w", err)
		}
		out = append(out, r)
	}
	return out, sc.Err()
}

// AppendCatalogRecord frames one record onto a stream for
// DecodeCatalogRecords.
func AppendCatalogRecord(stream []byte, id uint16, body []byte) []byte {
	stream = binary.LittleEndian.AppendUint16(stream, id)
	stream = binary.LittleEndian.AppendUint32(stream, uint32(len(body)))
	return append(stream, body...)
}

File: internal/wal3/rust/src/catalog.rs
//! Catalog/v3 record bodies (InfluxDB 3.10+ binary catalog), bitcode DERIVE
//! mode — a different encoding than the WAL's serde mode.
//!
//! Structs are vendored VERBATIM (field order, enum variant order, widths)
//! from influxdb3_catalog/src/format/records/*.rs; the shape is identical
//! across all upstream release tags v3.10.0..v3.11.4. Wire structs carry raw
//! ints — the upstream ID newtypes are in-memory only. Field order is
//! load-bearing (upstream freezes it with assert_roundtrip! hex snapshots,
//! which our tests mirror).
//!
//! Only the record types needed to answer "names, series keys, column ids"
//! are decoded; everything else is skipped by the Go caller via the record
//! length. Enums are positional: DO NOT reorder variants.

use serde::Serialize;

#[derive(bitcode::Decode, Serialize)]
pub struct CreateDatabase {
    pub database_id: u32,
    pub database_name: String,
    pub retention_period: RetentionPeriod,
}

#[derive(bitcode::Decode, Serialize)]
pub struct SoftDeleteDatabase {
    pub database_id: u32,
    pub deletion_time_ns: i64,
    pub hard_deletion_time_ns: Option<i64>,
    pub hard_delete_scope: Option<DeletionScope>,
}

#[derive(bitcode::Decode, Serialize)]
pub struct HardDeleteDatabase {
    pub db_id: u32,
}

#[derive(bitcode::Decode, Serialize)]
pub struct CreateTable {
    pub database_id: u32,
    pub database_name: String,
    pub table_name: String,
    pub table_id: u32,
    pub retention_period: RetentionPeriod,
    pub field_family_mode: FieldFamilyMode,
}

#[derive(bitcode::Decode, Serialize)]
pub struct SoftDeleteTable {
    pub database_id: u32,
    pub table_id: u32,
    pub deletion_time_ns: i64,
    pub hard_deletion_time_ns: Option<i64>,
    pub hard_delete_scope: Option<DeletionScope>,
}

#[derive(bitcode::Decode, Serialize)]
pub struct HardDeleteTable {
    pub db_id: u32,
    pub table_id: u32,
}

#[derive(bitcode::Decode, Serialize)]
pub struct AddColumns {
    pub database_id: u32,
    pub table_id: u32,
    pub columns: Vec<ColumnDefinition>,
    pub field_families: Vec<FieldFamilyDefinition>,
}

#[derive(bitcode::Decode, Serialize)]
pub enum ColumnDefinition {
    Timestamp(TimestampColumn), // 0
    Tag(TagColumn),             // 1
    Field(FieldColumn),         // 2
}

#[derive(bitcode::Decode, Serialize)]
pub struct TimestampColumn {
    pub column_id: Option<u16>,
    pub name: String,
}

#[derive(bitcode::Decode, Serialize)]
pub struct TagColumn {
    pub id: u16,
    pub column_id: Option<u16>,
    pub name: String,
}

#[derive(bitcode::Decode, Serialize)]
pub struct FieldColumn {
    pub id: FieldIdentifier,
    pub column_id: Option<u16>,
    pub name: String,
    pub data_type: FieldDataType,
}

#[derive(bitcode::Decode, Serialize)]
pub struct FieldIdentifier {
    pub family_id: u16,
    pub field_id: u16,
}

#[derive(bitcode::Decode, Serialize)]
pub enum FieldDataType {
    String,   // 0
    Integer,  // 1
    UInteger, // 2
    Float,    // 3
    Boolean,  // 4
    Binary,   // 5
}

#[derive(bitcode::Decode, Serialize)]
pub enum FieldFamilyName {
    User(String), // 0
    Auto(u16),    // 1
}

#[derive(bitcode::Decode, Serialize)]
pub struct FieldFamilyDefinition {
    pub id: u16,
    pub name: FieldFamilyName,
}

#[derive(bitcode::Decode, Serialize)]
pub enum FieldFamilyMode {
    Aware, // 0
    Auto,  // 1
}

#[derive(bitcode::Decode, Serialize)]
pub enum RetentionPeriod {
    Indefinite,                     // 0
    Duration { duration_secs: u64 }, // 1
}

#[derive(bitcode::Decode, Serialize)]
pub enum DeletionScope {
    DataAndCatalog,           // 0
    DataOnlyRemoveTables,     // 1
    DataOnlyKeepResources,    // 2
}

/// Record type ids (core partition; bit 15 set = enterprise partition).
pub const ID_CREATE_DATABASE: u16 = 4;
pub const ID_SOFT_DELETE_DATABASE: u16 = 5;
pub const ID_CREATE_TABLE: u16 = 6;
pub const ID_SOFT_DELETE_TABLE: u16 = 7;
pub const ID_ADD_COLUMNS: u16 = 8;
pub const ID_HARD_DELETE_DATABASE: u16 = 22;
pub const ID_HARD_DELETE_TABLE: u16 = 23;
pub const ID_RESTORE_CATALOG: u16 = 0x8006;

/// Decodes one record body to a JSON value string, or None for record types
/// the migration does not need (skipped upstream by length).
pub fn decode_record(id: u16, body: &[u8]) -> Result<Option<String>, String> {
    macro_rules! dec {
        ($ty:ty) => {{
            let v: $ty = bitcode::decode(body).map_err(|e| format!("record id {id}: {e}"))?;
            Some(serde_json::to_string(&v).map_err(|e| e.to_string())?)
        }};
    }
    Ok(match id {
        ID_CREATE_DATABASE => dec!(CreateDatabase),
        ID_SOFT_DELETE_DATABASE => dec!(SoftDeleteDatabase),
        ID_CREATE_TABLE => dec!(CreateTable),
        ID_SOFT_DELETE_TABLE => dec!(SoftDeleteTable),
        ID_ADD_COLUMNS => dec!(AddColumns),
        ID_HARD_DELETE_DATABASE => dec!(HardDeleteDatabase),
        ID_HARD_DELETE_TABLE => dec!(HardDeleteTable),
        ID_RESTORE_CATALOG => {
            return Err(format!(
                "record id {id} is an Enterprise catalog restore; state was swapped from backup files and cannot be replayed"
            ))
        }
        _ => None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hex(s: &str) -> Vec<u8> {
        (0..s.len()).step_by(2).map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap()).collect()
    }

    // Ground-truth byte vectors from upstream's own assert_roundtrip! tests
    // (influxdb3_catalog/src/format/records/*/tests.rs @ v3.10.0). If any of
    // these fail, the vendored shape or bitcode version drifted.
    #[test]
    fn create_database_duration() {
        let v: CreateDatabase = bitcode::decode(&hex("0401046d7964620104100e")).unwrap();
        assert_eq!(v.database_id, 1);
        assert_eq!(v.database_name, "mydb");
        assert!(matches!(v.retention_period, RetentionPeriod::Duration { duration_secs: 3600 }));
    }

    #[test]
    fn create_database_indefinite() {
        let v: CreateDatabase = bitcode::decode(&hex("042a07746573745f646200")).unwrap();
        assert_eq!(v.database_id, 42);
        assert_eq!(v.database_name, "test_db");
        assert!(matches!(v.retention_period, RetentionPeriod::Indefinite));
    }

    #[test]
    fn create_table() {
        let v: CreateTable = bitcode::decode(&hex("0401046d79646203637075040a01028051010001")).unwrap();
        assert_eq!((v.database_id, v.table_id), (1, 10));
        assert_eq!((v.database_name.as_str(), v.table_name.as_str()), ("mydb", "cpu"));
        assert!(matches!(v.retention_period, RetentionPeriod::Duration { duration_secs: 86400 }));
        assert!(matches!(v.field_family_mode, FieldFamilyMode::Auto));
    }

    #[test]
    fn hard_delete_table() {
        let v: HardDeleteTable = bitcode::decode(&hex("0401040a")).unwrap();
        assert_eq!((v.db_id, v.table_id), (1, 10));
    }

    #[test]
    fn add_columns_fields() {
        let v: AddColumns = bitcode::decode(&hex(
            "0401040a0208020000020102010100050876616c75656f766572666c6f770900",
        ))
        .unwrap();
        assert_eq!((v.database_id, v.table_id), (1, 10));
        assert_eq!(v.columns.len(), 2);
        match &v.columns[0] {
            ColumnDefinition::Field(f) => {
                assert_eq!(f.name, "value");
                assert_eq!(f.column_id, Some(1));
                assert!(matches!(f.data_type, FieldDataType::Float));
            }
            _ => panic!("expected field"),
        }
        match &v.columns[1] {
            ColumnDefinition::Field(f) => {
                assert_eq!(f.name, "overflow");
                assert_eq!(f.column_id, None);
                assert!(matches!(f.data_type, FieldDataType::Integer));
            }
            _ => panic!("expected field"),
        }
    }
}
